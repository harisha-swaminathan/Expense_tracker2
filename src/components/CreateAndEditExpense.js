import React, { useState } from "react";
import DatePicker from "react-datepicker";

const CreateAndEditExpense = (props) => {
    const expenseToEdit = props.expenseToEdit || {};    
    const [description, setDescription] =  useState(expenseToEdit.description ? expenseToEdit.description : '');
    const [amount, setAmount] =  useState(expenseToEdit.amount ?  (expenseToEdit.amount/100).toString() : '');
    const [note, setNote] =  useState(expenseToEdit.note ? expenseToEdit.note : '');
    const [createdAt, setCreatedAt] =  useState(expenseToEdit.createdAt || new Date());
    const [error, setError] =  useState('');

    const onDescriptionChange = (e) => {
        setDescription(e.target.value)
    };
    const onAmountChange = (e) => {
        const amount = e.target.value;

        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
            setAmount(e.target.value);
        }
    };
    const onNoteChange = (e) => {
        setNote(e.target.value)
    };
    const onDateChange = (createdAt) => {
        if(createdAt) {
            setCreatedAt(createdAt);
        }
    };
    const onExpenseFormSubmit = (e) => {
        e.preventDefault();
        if(!description || !amount) {
            setError('Please provide Description and amount!');
            return;
        }
        setError('');
        props.onSubmit({
            description,
            amount: parseFloat(amount, 10) * 100,
            createdAt: createdAt.valueOf(),
            note
        });
    };
    return (
        <>
            {!!error && <p>{error}</p>}
            <form onSubmit={onExpenseFormSubmit} data-testid="form-element">
                <input
                    type="text"
                    placeholder="Description"
                    onChange={onDescriptionChange}
                    value = {description}
                    autoFocus
                />
                <input
                    type="text"
                    placeholder="Amount"
                    onChange={onAmountChange}
                    value = {amount}
                />
                <DatePicker 
                    selected={createdAt} 
                    showIcon
                    format="dd/MM/yy"
                    onChange={onDateChange} 
                />
                <textarea
                    placeholder="add a note for you expense"
                    onChange={onNoteChange}
                    value = {note}
                />
                <button data-testid="form-submit"> Add Expense</button>
            </form>
        </>
    );
}

export default CreateAndEditExpense;

