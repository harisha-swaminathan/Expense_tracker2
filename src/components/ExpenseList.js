import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from "../selectors/expenses";

export const ExpenseList = (props) => {
    return (
        <>
            {
                props.expenses.length ===0 ? (
                    <p> No expenses </p>
                ) : (
                    props.expenses.map((expense) => {
                        return <ExpenseListItem key={expense.id} {...expense}/>
                    })
                )}
        </>
    );
}
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
};

const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);

export default ConnectedExpenseList;