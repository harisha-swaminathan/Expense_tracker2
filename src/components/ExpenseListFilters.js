import React, { useState } from 'react';
import {connect} from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setEndDate, setStartDate} from '../actions/filters';
import DatePicker from "react-datepicker";

export const ExpenseListFilter = (props) => {
    const startDate = props.filters.startDate ? props.filters.startDate.toString() : null;
    const endDate = props.filters.endDate ? props.filters.endDate.toString() : null;
    const text = props.filters.text;
    return (
        <>
        <input 
            placeholder='filter by text'
            type="text"
            value={text}
            onChange={(e) => {
                props.setTextFilter(e.target.value)
            }
        }/>
        <select
            data-testid="select-filter"
            value="date"
            onChange={(e) => {
                if(e.target.value === "amount"){
                    props.sortByAmount();
                } else {
                    props.sortByDate();
                }
            }
        }>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
        </select>
        <DatePicker
            selected={startDate}
            showIcon
            onChange={(date) => {
                props.setStartDate(date)
            }}
            selectsStart
            startDate={startDate}
            endDate={endDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        <DatePicker
            selected={endDate}
            showIcon
            onChange={(date) => {
                props.setEndDate(date)
            }}
            selectsEnd
            startDate={startDate}
            endDate={endDate}
            minDate={startDate}
            dateFormat="MM/dd/yyyy"
            isClearable={true}
        />
        </>
    );
};

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
};
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => {
        dispatch(setTextFilter(value));
    },
    sortByAmount: () => {
        dispatch(sortByAmount());
    },
    sortByDate: () => {
        dispatch(sortByDate());
    },
    setStartDate: (date) => {
        dispatch(setStartDate(date));
    },
    setEndDate: (date) => {
        dispatch(setEndDate(date));
    }
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);