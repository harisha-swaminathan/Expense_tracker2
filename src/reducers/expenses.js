const expenseReducerDefaultState = [];

const expensesReducer = (state = expenseReducerDefaultState, action) => {
    switch(action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense];
        case 'DELETE_EXPENSE':
            return state.filter((expense) => expense.id !== action.id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            })
        default:
            return state;
    }
};

export default expensesReducer;