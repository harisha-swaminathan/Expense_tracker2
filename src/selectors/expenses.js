import { isBefore, isAfter } from "date-fns";


const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
      
        const description = expense.description.toLowerCase();
        const textMatch = typeof text !== 'string' || description.includes(text.toLowerCase());
        const startDateMatch = !startDate || isAfter(expense.createdAt, startDate); 
        const endDateMatch =  !endDate || isBefore(expense.createdAt, endDate);

        return textMatch && startDateMatch && endDateMatch;
    }).sort((expense1, expense2) => {
        if(sortBy == 'date'){
            return expense1.createdAt < expense2.createdAt ? 1 : -1;
        } else {
            return expense1.amount < expense2.amount ? 1 : -1;
        }
    })
};

export default getVisibleExpenses;