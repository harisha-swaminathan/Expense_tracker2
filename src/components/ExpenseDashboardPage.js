import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilters';

const ExpensesDashboardPage = () => (
    <div>
        
        <ExpenseListFilter/>
        <ExpenseList/>
    </div>
);
export default ExpensesDashboardPage;