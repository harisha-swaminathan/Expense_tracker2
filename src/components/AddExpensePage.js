import React from 'react';
import CreateAndEditExpense from './CreateAndEditExpense';
import { connect } from "react-redux";
import { addExpense } from "../actions/expenses";
import { useNavigate } from "react-router";

export const AddExpensePage = (props) => {
    const navigator = useNavigate();
    return (
        <div>
            <h1> Add Expense</h1>
            <CreateAndEditExpense 
                onSubmit={(expense) => {
                   props.addExpense(expense)
                    navigator("/");
                }}
            />
        </div>
    );
}

const mapDispatchToProps = (dispatch) => ({
    
    addExpense: ({description, note, amount, createdAt}) => {
        dispatch(addExpense({ 
            description,
            note,
            amount,
            createdAt
        }));
    }
    
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);