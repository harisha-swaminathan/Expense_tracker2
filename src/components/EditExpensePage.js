import React from 'react';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import CreateAndEditExpense from './CreateAndEditExpense';
import { editExpense , deleteExpense} from '../actions/expenses';
import { useNavigate } from "react-router";

export const EditExpensePage = (props) => {
    const navigator = useNavigate();
    const id = useParams().id;
    const expenseToEdit = props.expenses.find((expense) => expense.id === id);
    const onEditExpense = (updates) => {
        props.editExpense(id, updates);
        navigator("/");
    };
    const onRemove = () => {
        props.deleteExpense(id);
        navigator("/");
    };
    return (
        <div>
            <CreateAndEditExpense expenseToEdit={expenseToEdit} onSubmit={(onEditExpense)}/>
            <button onClick={onRemove}> Remove </button>
        </div>
    )
};

const mapStateToProps = (state) => {
    return {
        expenses: state.expenses
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        deleteExpense: (id) => {
            dispatch(deleteExpense(id))
        },
        editExpense: (id, updates) => {
            dispatch(editExpense(id, updates))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);