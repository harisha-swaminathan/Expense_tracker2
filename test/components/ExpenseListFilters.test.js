import React from 'react';
import { render, screen, fireEvent} from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { ExpenseListFilter } from "../../src/components/ExpenseListFilters";
import { BrowserRouter } from 'react-router-dom';
import expenses from '../fixtures/expenses';
import filters from '../fixtures/filters';
import '@testing-library/jest-dom';

let setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate;

beforeEach(() => {
    setTextFilter = jest.fn();
    sortByDate = jest.fn();
    sortByAmount = jest.fn();
    setStartDate = jest.fn();
    setEndDate = jest.fn();

});

test('render expense list filter', () => {
    const { asFragment } = render(<ExpenseListFilter 
        filters={filters[0]}
        setTextFilter = {setTextFilter}
        sortByAmount = {sortByAmount}
        sortByDate = {sortByDate}
        setStartDate = {setStartDate}
        setEndDate = {setEndDate}
        />
    );
    expect(asFragment()).toMatchSnapshot();
});

test('render expense list second filter', () => {
    const { asFragment } = render(<ExpenseListFilter 
        filters={filters[1]}
        setTextFilter = {setTextFilter}
        sortByAmount = {sortByAmount}
        sortByDate = {sortByDate}
        setStartDate = {setStartDate}
        setEndDate = {setEndDate}
        />
    );
    expect(asFragment()).toMatchSnapshot();
});

test('handle text change', () => {
    render(<ExpenseListFilter 
        filters={filters[1]}
        setTextFilter = {setTextFilter}
        sortByAmount = {sortByAmount}
        sortByDate = {sortByDate}
        setStartDate = {setStartDate}
        setEndDate = {setEndDate}
        />
    );
    const textFilter = screen.getByPlaceholderText('filter by text');

    fireEvent.change(textFilter, {target: {value: 'rent'}});
    expect(setTextFilter).toBeCalledWith('rent');
});

test('should sort by date', () => {
    render(<ExpenseListFilter 
        filters={filters[1]}
        setTextFilter = {setTextFilter}
        sortByAmount = {sortByAmount}
        sortByDate = {sortByDate}
        setStartDate = {setStartDate}
        setEndDate = {setEndDate}
        />
    );
    const sortByDateInput = screen.getByTestId("select-filter");

    fireEvent.change(sortByDateInput,  { target: { value: 'date' } })
    expect(sortByDate).toHaveBeenCalled();
});

test('should sort by amount', () => {
    render(<ExpenseListFilter 
        filters={filters[1]}
        setTextFilter = {setTextFilter}
        sortByAmount = {sortByAmount}
        sortByDate = {sortByDate}
        setStartDate = {setStartDate}
        setEndDate = {setEndDate}
        />
    );
    const sortByAmountInput = screen.getByTestId("select-filter");

    fireEvent.change(sortByAmountInput,  { target: { value: 'amount' } })
    expect(sortByAmount).toHaveBeenCalled();
});