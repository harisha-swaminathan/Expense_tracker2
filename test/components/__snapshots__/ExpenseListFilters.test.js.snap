// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render expense list filter 1`] = `
<DocumentFragment>
  <input
    placeholder="filter by text"
    type="text"
    value=""
  />
  <select
    data-testid="select-filter"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container react-datepicker__view-calendar-icon"
    >
      <svg
        class="react-datepicker__calendar-icon "
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
        />
      </svg>
      <input
        class=""
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container react-datepicker__view-calendar-icon"
    >
      <svg
        class="react-datepicker__calendar-icon "
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
        />
      </svg>
      <input
        class=""
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`render expense list second filter 1`] = `
<DocumentFragment>
  <input
    placeholder="filter by text"
    type="text"
    value="bill"
  />
  <select
    data-testid="select-filter"
  >
    <option
      value="date"
    >
      Date
    </option>
    <option
      value="amount"
    >
      Amount
    </option>
  </select>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container react-datepicker__view-calendar-icon"
    >
      <svg
        class="react-datepicker__calendar-icon "
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
        />
      </svg>
      <input
        class=""
        type="text"
        value="12/31/1969"
      />
      <button
        aria-label="Close"
        class="react-datepicker__close-icon"
        tabindex="-1"
        type="button"
      />
    </div>
  </div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container react-datepicker__view-calendar-icon"
    >
      <svg
        class="react-datepicker__calendar-icon "
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M96 32V64H48C21.5 64 0 85.5 0 112v48H448V112c0-26.5-21.5-48-48-48H352V32c0-17.7-14.3-32-32-32s-32 14.3-32 32V64H160V32c0-17.7-14.3-32-32-32S96 14.3 96 32zM448 192H0V464c0 26.5 21.5 48 48 48H400c26.5 0 48-21.5 48-48V192z"
        />
      </svg>
      <input
        class=""
        type="text"
        value=""
      />
      <button
        aria-label="Close"
        class="react-datepicker__close-icon"
        tabindex="-1"
        type="button"
      />
    </div>
  </div>
</DocumentFragment>
`;
